### PR TITLE
FIX: make sure we have more than 1 tick with small log ranges

### LINKED ISSUE #8768: One one tick in a log-scale axis
### Bug report

For the data in a small range, there is just on tick in the default log scale axis, which means nothing.

**Code for reproduction**

```python
import matplotlib.pylab as plt
plt.loglog([ 0.53,  0.41],[ 0.53,  0.41] )
plt.show()
```

**Actual outcome**

image

**Expected outcome**
Give at least two ticks for each axis.

**Matplotlib version**

  * Operating System: Ubuntu 16.04
  * Matplotlib Version: 2.02
  * Python Version: 3.6.0
I install the library by conda.

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

If we have both a small target number of ticks and a small displayed
range we would previously pick a tick stride that was the same size or
bigger than the visible data range which would result in only 1 tick
being visible on the axis.

This patch ensures that, with a floor at 1, the stride is smaller than
the estimated data range.

re-closes #8768  (same symptoms, different cause)

attn @aggna 

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).

### CODE REVIEW

File: lib/matplotlib/ticker.py [hunk 1]
```diff
@@ -2479,6 +2479,13 @@ def tick_values(self, vmin, vmax):
                   if mpl.rcParams['_internal.classic_mode'] else
                   (numdec + 1) // numticks + 1)
 
+        # if we have decided that the stride is as big or bigger than
```
Comment (jklymak) [inline lib/matplotlib/ticker.py:2482]: Is this really the right solution to the original problem?  To me, it seems that `numticks` is 1 because there isn't room for more ticks.  Do we really want ticks that overwrite each other for very narrow axes?

Comment (timhoffm) [reply lib/matplotlib/ticker.py:2482]: If you are very narrow, there's no good solution. Either you have only one tick, which makes your scale uninterpretable. Or you have overlap.

Comment (jklymak) [reply lib/matplotlib/ticker.py:2482]: Sorry, let me rephrase: We clip numticks to 2, so why don't we have 2 ticks?

Comment (tacaswell) [reply lib/matplotlib/ticker.py:2482]: `numticks` is an upper bound and only "best effort" at that.  

The issue is that based on `numtick` and a rounded "in" version of the range we pick a stride between major ticks (which is meant to catch things like 30 orders of magnitude with 9 ticks).  If we propose 3 ticks, have a range of [5, 150], then we "round out" to [1, 100] for the range giving us `numdec = 2` and `numticks = 3` and a stride of `(3 / 3) + 1 == 2` which means we put ticks at `(10 ^ {-2, 0, 2, 4})` which in turn means that only the tick at 100 is actually shown.  Additionally because we only show the minor ticks if the major stride is equal to 1 you end up with exactly 1 tick and 1 label shown even though the users / auto logic asked for up to 3.

I think that this fix (which fixes the values after doing a bunch of heuristics) is a better and more reliable fix than making the heuristics more complicated.

**Review by jklymak — APPROVED** [review]
Does this fix the minor tick issue?

**Comment by tacaswell** [comment]
> Does this fix the minor tick issue?

Mostly.  The issue I described can be pushed up a bit to hit the case where we should have ~2 decades in view and we pick a 3 decade stride and in that case we still don't have the minor ticks, but will at least get 2 major major ticks.

I did some manually testing with this by making the y-axis super small and then interactively panning / zooming to see if anything bad looking happened...

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/tests/test_ticker.py [hunk 2]
```diff
@@ -1362,3 +1362,13 @@ def test_bad_locator_subs(sub):
...
+
+
+@pytest.mark.parametrize('numticks', [1, 2, 3, 9])
+@pytest.mark.style('default')
+def test_small_range_loglocator(numticks):
+    ll = mticker.LogLocator()
+    ll.set_params(numticks=numticks)
+    for top in [5, 7, 9, 11, 15, 50, 100, 1000]:
+        ticks = ll.tick_values(.5, top)
+        assert (np.diff(np.log10(ll.tick_values(6, 150))) == 1).all()
```
Comment (anntzer) [inline lib/matplotlib/tests/test_ticker.py:1374]: @tacaswell There seems to be a clear typo in the test here: you don't do anything with `ticks`, and the assert is the same on each loop iteration.  Do you remember what you intended to do?
